File: js/genkit/src/datasets.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { mkdir, readFile, rm, writeFile } from 'node:fs/promises';
import path from 'node:path';
import { randomUUID } from 'node:crypto';

/** The target surface a dataset is intended for. */
export type DatasetType =
  | 'UNKNOWN'
  | 'FLOW'
  | 'MODEL'
  | 'EXECUTABLE_PROMPT';

/** JSON schema hints for dataset `input` and `reference` fields. */
export interface DatasetSchema {
  inputSchema?: Record<string, unknown>;
  referenceSchema?: Record<string, unknown>;
}

/** User-facing sample shape accepted when creating or updating a dataset. */
export interface InferenceSample {
  testCaseId?: string;
  input: unknown;
  reference?: unknown;
}

/** Persisted dataset sample shape used by the developer UI. */
export interface DatasetSample {
  testCaseId: string;
  input: unknown;
  reference?: unknown;
}

/** Request payload for creating a dataset. */
export interface CreateDatasetRequest {
  data: InferenceSample[];
  datasetId?: string;
  datasetType: DatasetType;
  schema?: DatasetSchema;
  metricRefs?: string[];
  targetAction?: string;
}

/** Request payload for updating a dataset. */
export interface UpdateDatasetRequest {
  datasetId: string;
  data?: InferenceSample[];
  schema?: DatasetSchema;
  metricRefs?: string[];
  targetAction?: string;
}

/** Metadata shown for a dataset in the developer UI. */
export interface DatasetMetadata {
  datasetId: string;
  size: number;
  schema?: DatasetSchema;
  datasetType: DatasetType;
  targetAction?: string;
  metricRefs: string[];
  version: number;
  createTime: string;
  updateTime: string;
}

/**
 * DatasetStore is the narrow persistence boundary for eval datasets.
 *
 * The current developer UI expects the local `.genkit/datasets` file layout,
 * but callers should depend on this interface so a remote datastore can be
 * introduced later without changing the programmatic API.
 */
export interface DatasetStore {
  createDataset(req: CreateDatasetRequest): Promise<DatasetMetadata>;
  updateDataset(req: UpdateDatasetRequest): Promise<DatasetMetadata>;
  getDataset(datasetId: string): Promise<DatasetSample[]>;
  listDatasets(): Promise<DatasetMetadata[]>;
  deleteDataset(datasetId: string): Promise<void>;
}

const DATASET_ID_RE = /^[A-Za-z][A-Za-z0-9_.-]{4,34}[A-Za-z0-9]$/;

/**
 * LocalFileDatasetStore is the default DatasetStore implementation used by the
 * current developer UI. It persists `index.json` plus one `${datasetId}.json`
 * file under `.genkit/datasets`.
 */
export class LocalFileDatasetStore implements DatasetStore {
  readonly storeRoot: string;
  readonly indexFile: string;

  /** Creates a store rooted at a dataset directory. */
  constructor(storeRoot: string) {
    this.storeRoot = storeRoot;
    this.indexFile = path.join(storeRoot, 'index.json');
  }

  /** Returns the default local dataset store for a project root. */
  static async forProjectRoot(projectRoot = process.cwd()) {
    const store = new LocalFileDatasetStore(
      path.join(projectRoot, '.genkit', 'datasets')
    );
    await store.ensureInitialized();
    return store;
  }

  /** Creates a dataset using the current developer UI file format. */
  async createDataset(req: CreateDatasetRequest): Promise<DatasetMetadata> {
    await this.ensureInitialized();
    const metadataMap = await this.readMetadataMap();
    const datasetId = this.generateDatasetId(req.datasetId, metadataMap);
    const dataset = normalizeInferenceDataset(req.data);

    await writeFile(
      this.datasetFilePath(datasetId),
      JSON.stringify(dataset),
      'utf8'
    );

    const now = new Date().toString();
    const metadata: DatasetMetadata = {
      datasetId,
      size: dataset.length,
      schema: req.schema,
      datasetType: req.datasetType,
      targetAction: req.targetAction,
      metricRefs: [...(req.metricRefs ?? [])],
      version: 1,
      createTime: now,
      updateTime: now,
    };
    metadataMap[datasetId] = metadata;
    await this.writeMetadataMap(metadataMap);
    return metadata;
  }

  /** Updates a dataset and increments its version when sample data changes. */
  async updateDataset(req: UpdateDatasetRequest): Promise<DatasetMetadata> {
    await this.ensureInitialized();
    const filePath = this.datasetFilePath(req.datasetId);
    const metadataMap = await this.readMetadataMap();
    const prevMetadata = metadataMap[req.datasetId];
    if (!prevMetadata) {
      throw new Error('Update dataset failed: dataset metadata not found');
    }

    let newSize = prevMetadata.size;
    let version = prevMetadata.version;
    if (req.data) {
      const patched = await this.patchDataset(
        req.datasetId,
        normalizeInferenceDataset(req.data)
      );
      newSize = patched.length;
      version += 1;
    } else {
      await readFile(filePath, 'utf8');
    }

    const metadata: DatasetMetadata = {
      datasetId: req.datasetId,
      size: newSize,
      schema: req.schema ?? prevMetadata.schema,
      datasetType: prevMetadata.datasetType,
      targetAction: req.targetAction ?? prevMetadata.targetAction,
      metricRefs: req.metricRefs ? [...req.metricRefs] : [...prevMetadata.metricRefs],
      version,
      createTime: prevMetadata.createTime,
      updateTime: new Date().toString(),
    };
    metadataMap[req.datasetId] = metadata;
    await this.writeMetadataMap(metadataMap);
    return metadata;
  }

  /** Reads the persisted samples for a dataset ID. */
  async getDataset(datasetId: string): Promise<DatasetSample[]> {
    const raw = await readFile(this.datasetFilePath(datasetId), 'utf8');
    return JSON.parse(raw) as DatasetSample[];
  }

  /** Lists all dataset metadata records sorted by dataset ID. */
  async listDatasets(): Promise<DatasetMetadata[]> {
    const metadataMap = await this.readMetadataMap();
    return Object.keys(metadataMap)
      .sort()
      .map((datasetId) => metadataMap[datasetId]);
  }

  /** Deletes a dataset file and its metadata entry. */
  async deleteDataset(datasetId: string): Promise<void> {
    await rm(this.datasetFilePath(datasetId));
    const metadataMap = await this.readMetadataMap();
    delete metadataMap[datasetId];
    await this.writeMetadataMap(metadataMap);
  }

  private async ensureInitialized() {
    await mkdir(this.storeRoot, { recursive: true });
    try {
      await readFile(this.indexFile, 'utf8');
    } catch {
      await writeFile(this.indexFile, JSON.stringify({}), 'utf8');
    }
  }

  private async readMetadataMap(): Promise<Record<string, DatasetMetadata>> {
    const raw = await readFile(this.indexFile, 'utf8');
    return JSON.parse(raw) as Record<string, DatasetMetadata>;
  }

  private async writeMetadataMap(
    metadataMap: Record<string, DatasetMetadata>
  ): Promise<void> {
    await writeFile(this.indexFile, JSON.stringify(metadataMap), 'utf8');
  }

  private async patchDataset(
    datasetId: string,
    patch: DatasetSample[]
  ): Promise<DatasetSample[]> {
    const existing = await this.getDataset(datasetId);
    const datasetMap = new Map(existing.map((item) => [item.testCaseId, item]));
    for (const sample of patch) {
      datasetMap.set(sample.testCaseId, sample);
    }
    const merged = Array.from(datasetMap.values());
    await writeFile(
      this.datasetFilePath(datasetId),
      JSON.stringify(merged),
      'utf8'
    );
    return merged;
  }

  private generateDatasetId(
    datasetId: string | undefined,
    metadataMap: Record<string, DatasetMetadata>
  ) {
    if (!datasetId) {
      let generated = randomUUID();
      while (metadataMap[generated]) {
        generated = randomUUID();
      }
      return generated;
    }
    if (!DATASET_ID_RE.test(datasetId)) {
      throw new Error(
        'Invalid datasetId provided. ID must be alphanumeric, with hyphens, dots and dashes. Is must start with an alphabet, end with an alphabet or a number, and must be 6-36 characters long.'
      );
    }
    if (metadataMap[datasetId]) {
      throw new Error(`Dataset ID not unique: ${datasetId}`);
    }
    return datasetId;
  }

  private datasetFilePath(datasetId: string) {
    return path.join(this.storeRoot, `${datasetId}.json`);
  }
}

/**
 * Returns the default DatasetStore for the current project.
 *
 * Today this is file-backed so datasets render in the current developer UI.
 * Callers should keep using the DatasetStore interface so a remote
 * implementation can be swapped in later if needed.
 */
export async function datasetStoreForProjectRoot(
  projectRoot = process.cwd()
): Promise<DatasetStore> {
  return LocalFileDatasetStore.forProjectRoot(projectRoot);
}

function normalizeInferenceDataset(data: InferenceSample[]): DatasetSample[] {
  return data.map((sample) => ({
    testCaseId: sample.testCaseId ?? randomUUID(),
    input: sample.input,
    reference: sample.reference,
  }));
}

File: go/genkit/datasets.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package genkit

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"slices"
	"time"

	"github.com/google/uuid"
)

var datasetIDPattern = regexp.MustCompile(`^[A-Za-z][A-Za-z0-9_.-]{4,34}[A-Za-z0-9]$`)

// DatasetType identifies the target surface the dataset is intended for.
type DatasetType string

const (
	DatasetTypeUnknown          DatasetType = "UNKNOWN"
	DatasetTypeFlow             DatasetType = "FLOW"
	DatasetTypeModel            DatasetType = "MODEL"
	DatasetTypeExecutablePrompt DatasetType = "EXECUTABLE_PROMPT"
)

// DatasetSchema describes the shape of the dataset's input and reference values.
type DatasetSchema struct {
	InputSchema     map[string]any `json:"inputSchema,omitempty"`
	ReferenceSchema map[string]any `json:"referenceSchema,omitempty"`
}

// InferenceSample is the user-facing dataset item shape.
type InferenceSample struct {
	TestCaseID string `json:"testCaseId,omitempty"`
	Input      any    `json:"input"`
	Reference  any    `json:"reference,omitempty"`
}

// DatasetSample is the persisted dataset item shape used by the developer UI.
type DatasetSample struct {
	TestCaseID string `json:"testCaseId"`
	Input      any    `json:"input"`
	Reference  any    `json:"reference,omitempty"`
}

// CreateDatasetRequest defines the payload for creating a dataset.
type CreateDatasetRequest struct {
	Data         []InferenceSample `json:"data"`
	DatasetID    string            `json:"datasetId,omitempty"`
	DatasetType  DatasetType       `json:"datasetType"`
	Schema       *DatasetSchema    `json:"schema,omitempty"`
	MetricRefs   []string          `json:"metricRefs,omitempty"`
	TargetAction string            `json:"targetAction,omitempty"`
}

// UpdateDatasetRequest defines the payload for updating a dataset.
type UpdateDatasetRequest struct {
	DatasetID    string            `json:"datasetId"`
	Data         []InferenceSample `json:"data,omitempty"`
	Schema       *DatasetSchema    `json:"schema,omitempty"`
	MetricRefs   []string          `json:"metricRefs,omitempty"`
	TargetAction string            `json:"targetAction,omitempty"`
}

// DatasetMetadata contains the dataset metadata shown in the developer UI.
type DatasetMetadata struct {
	DatasetID    string         `json:"datasetId"`
	Size         int            `json:"size"`
	Schema       *DatasetSchema `json:"schema,omitempty"`
	DatasetType  DatasetType    `json:"datasetType"`
	TargetAction string         `json:"targetAction,omitempty"`
	MetricRefs   []string       `json:"metricRefs"`
	Version      int            `json:"version"`
	CreateTime   string         `json:"createTime"`
	UpdateTime   string         `json:"updateTime"`
}

// LocalFileDatasetStore is the default DatasetStore implementation used by the
// current developer UI. It persists datasets as index.json plus one
// <datasetId>.json file under .genkit/datasets.
type LocalFileDatasetStore struct {
	storeRoot string
	indexFile string
	now       func() time.Time
}

// NewLocalFileDatasetStore creates a dataset store rooted at the provided path.
func NewLocalFileDatasetStore(storeRoot string) (*LocalFileDatasetStore, error) {
	if storeRoot == "" {
		return nil, errors.New("store root is required")
	}
	if err := os.MkdirAll(storeRoot, 0755); err != nil {
		return nil, fmt.Errorf("create dataset store root: %w", err)
	}

	store := &LocalFileDatasetStore{
		storeRoot: storeRoot,
		indexFile: filepath.Join(storeRoot, "index.json"),
		now:       time.Now,
	}

	if _, err := os.Stat(store.indexFile); errors.Is(err, os.ErrNotExist) {
		if err := store.writeMetadataMap(map[string]DatasetMetadata{}); err != nil {
			return nil, err
		}
	} else if err != nil {
		return nil, fmt.Errorf("stat dataset index: %w", err)
	}

	return store, nil
}

// DefaultDatasetStore creates the default dataset store for this project.
//
// Today this is a LocalFileDatasetStore so datasets render in the current
// developer UI. If Genkit adds remote dataset persistence later, introduce that
// abstraction at the consumer boundary rather than coupling callers to the
// current file layout.
func DefaultDatasetStore() (*LocalFileDatasetStore, error) {
	projectRoot, err := findProjectRoot()
	if err != nil {
		return nil, err
	}
	return NewLocalFileDatasetStore(filepath.Join(projectRoot, ".genkit", "datasets"))
}

// CreateDataset writes a new dataset and its metadata using the developer UI's
// current on-disk format.
func (s *LocalFileDatasetStore) CreateDataset(req CreateDatasetRequest) (*DatasetMetadata, error) {
	metadataMap, err := s.readMetadataMap()
	if err != nil {
		return nil, err
	}

	datasetID, err := s.generateDatasetID(req.DatasetID, metadataMap)
	if err != nil {
		return nil, err
	}

	dataset := normalizeInferenceDataset(req.Data)
	if err := writeJSONFile(s.datasetFilePath(datasetID), dataset); err != nil {
		return nil, fmt.Errorf("write dataset: %w", err)
	}

	now := s.now().String()
	metricRefs := cloneStrings(req.MetricRefs)
	metadata := DatasetMetadata{
		DatasetID:    datasetID,
		Size:         len(dataset),
		Schema:       req.Schema,
		DatasetType:  req.DatasetType,
		TargetAction: req.TargetAction,
		MetricRefs:   metricRefs,
		Version:      1,
		CreateTime:   now,
		UpdateTime:   now,
	}
	metadataMap[datasetID] = metadata

	if err := s.writeMetadataMap(metadataMap); err != nil {
		return nil, err
	}
	return &metadata, nil
}

// UpdateDataset merges the provided samples into an existing dataset and bumps
// the dataset version when sample data changes.
func (s *LocalFileDatasetStore) UpdateDataset(req UpdateDatasetRequest) (*DatasetMetadata, error) {
	filePath := s.datasetFilePath(req.DatasetID)
	if _, err := os.Stat(filePath); err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, errors.New("update dataset failed: dataset not found")
		}
		return nil, fmt.Errorf("stat dataset: %w", err)
	}

	metadataMap, err := s.readMetadataMap()
	if err != nil {
		return nil, err
	}
	prevMetadata, ok := metadataMap[req.DatasetID]
	if !ok {
		return nil, errors.New("update dataset failed: dataset metadata not found")
	}

	newSize := prevMetadata.Size
	version := prevMetadata.Version
	if req.Data != nil {
		patched, err := s.patchDataset(req.DatasetID, normalizeInferenceDataset(req.Data))
		if err != nil {
			return nil, err
		}
		newSize = len(patched)
		version++
	}

	newMetadata := DatasetMetadata{
		DatasetID:    req.DatasetID,
		Size:         newSize,
		Schema:       firstNonNil(req.Schema, prevMetadata.Schema),
		DatasetType:  prevMetadata.DatasetType,
		TargetAction: firstNonEmpty(req.TargetAction, prevMetadata.TargetAction),
		MetricRefs:   firstNonNilStrings(req.MetricRefs, prevMetadata.MetricRefs),
		Version:      version,
		CreateTime:   prevMetadata.CreateTime,
		UpdateTime:   s.now().String(),
	}
	metadataMap[req.DatasetID] = newMetadata

	if err := s.writeMetadataMap(metadataMap); err != nil {
		return nil, err
	}
	return &newMetadata, nil
}

// GetDataset reads the persisted samples for a dataset ID.
func (s *LocalFileDatasetStore) GetDataset(datasetID string) ([]DatasetSample, error) {
	var dataset []DatasetSample
	if err := readJSONFile(s.datasetFilePath(datasetID), &dataset); err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, fmt.Errorf("dataset not found for dataset ID %s", datasetID)
		}
		return nil, fmt.Errorf("read dataset: %w", err)
	}
	return dataset, nil
}

// ListDatasets returns all dataset metadata records sorted by dataset ID.
func (s *LocalFileDatasetStore) ListDatasets() ([]DatasetMetadata, error) {
	metadataMap, err := s.readMetadataMap()
	if err != nil {
		return nil, err
	}

	ids := make([]string, 0, len(metadataMap))
	for datasetID := range metadataMap {
		ids = append(ids, datasetID)
	}
	slices.Sort(ids)

	metadatas := make([]DatasetMetadata, 0, len(ids))
	for _, datasetID := range ids {
		metadatas = append(metadatas, metadataMap[datasetID])
	}
	return metadatas, nil
}

// DeleteDataset removes a dataset file and its metadata entry.
func (s *LocalFileDatasetStore) DeleteDataset(datasetID string) error {
	if err := os.Remove(s.datasetFilePath(datasetID)); err != nil {
		return fmt.Errorf("delete dataset: %w", err)
	}

	metadataMap, err := s.readMetadataMap()
	if err != nil {
		return err
	}
	delete(metadataMap, datasetID)
	return s.writeMetadataMap(metadataMap)
}

func (s *LocalFileDatasetStore) patchDataset(datasetID string, patch []DatasetSample) ([]DatasetSample, error) {
	existing, err := s.GetDataset(datasetID)
	if err != nil {
		return nil, err
	}

	byID := make(map[string]DatasetSample, len(existing))
	order := make([]string, 0, len(existing))
	for _, sample := range existing {
		byID[sample.TestCaseID] = sample
		order = append(order, sample.TestCaseID)
	}

	for _, sample := range patch {
		if _, exists := byID[sample.TestCaseID]; !exists {
			order = append(order, sample.TestCaseID)
		}
		byID[sample.TestCaseID] = sample
	}

	patched := make([]DatasetSample, 0, len(order))
	for _, id := range order {
		patched = append(patched, byID[id])
	}

	if err := writeJSONFile(s.datasetFilePath(datasetID), patched); err != nil {
		return nil, fmt.Errorf("write patched dataset: %w", err)
	}
	return patched, nil
}

func (s *LocalFileDatasetStore) readMetadataMap() (map[string]DatasetMetadata, error) {
	metadataMap := map[string]DatasetMetadata{}
	if err := readJSONFile(s.indexFile, &metadataMap); err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return metadataMap, nil
		}
		return nil, fmt.Errorf("read dataset index: %w", err)
	}
	return metadataMap, nil
}

func (s *LocalFileDatasetStore) writeMetadataMap(metadataMap map[string]DatasetMetadata) error {
	if metadataMap == nil {
		metadataMap = map[string]DatasetMetadata{}
	}
	if err := writeJSONFile(s.indexFile, metadataMap); err != nil {
		return fmt.Errorf("write dataset index: %w", err)
	}
	return nil
}

func (s *LocalFileDatasetStore) generateDatasetID(datasetID string, metadataMap map[string]DatasetMetadata) (string, error) {
	if datasetID == "" {
		for {
			candidate := uuid.New().String()
			if _, exists := metadataMap[candidate]; !exists {
				return candidate, nil
			}
		}
	}
	if !datasetIDPattern.MatchString(datasetID) {
		return "", errors.New("invalid datasetId provided. ID must be alphanumeric, with hyphens, dots and dashes. Is must start with an alphabet, end with an alphabet or a number, and must be 6-36 characters long.")
	}
	if _, exists := metadataMap[datasetID]; exists {
		return "", fmt.Errorf("dataset ID not unique: %s", datasetID)
	}
	return datasetID, nil
}

func (s *LocalFileDatasetStore) datasetFilePath(datasetID string) string {
	return filepath.Join(s.storeRoot, datasetID+".json")
}

func normalizeInferenceDataset(data []InferenceSample) []DatasetSample {
	dataset := make([]DatasetSample, 0, len(data))
	for _, sample := range data {
		testCaseID := sample.TestCaseID
		if testCaseID == "" {
			testCaseID = uuid.New().String()
		}
		dataset = append(dataset, DatasetSample{
			TestCaseID: testCaseID,
			Input:      sample.Input,
			Reference:  sample.Reference,
		})
	}
	return dataset
}

func writeJSONFile(path string, value any) error {
	bytes, err := json.Marshal(value)
	if err != nil {
		return err
	}
	return os.WriteFile(path, bytes, 0644)
}

func readJSONFile(path string, dest any) error {
	bytes, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	return json.Unmarshal(bytes, dest)
}

func firstNonEmpty(value string, fallback string) string {
	if value != "" {
		return value
	}
	return fallback
}

func firstNonNil[T any](value *T, fallback *T) *T {
	if value != nil {
		return value
	}
	return fallback
}

func firstNonNilStrings(value []string, fallback []string) []string {
	if value != nil {
		return cloneStrings(value)
	}
	return cloneStrings(fallback)
}

func cloneStrings(values []string) []string {
	if len(values) == 0 {
		return []string{}
	}
	return append([]string(nil), values...)
}
